apps/camera/js/lib/storage.js

```diff
@@ -223,15 +232,59 @@ Storage.prototype.configure = function(storageVolumeName) {
 
   // Shouldn't happen?
   if (!this.picture) {
+    this.watchStorages = [];
     this.setState('unavailable');
     return;
   }
 
+  // The picture storage state is managed separately
+  this.watchStorages = picturesStorages;
+  for (i = 0; i < this.watchStorages.length;) {
+    if (!this.watchStorages[i] || this.watchStorages[i] === this.picture) {
+      this.watchStorages.splice(i, 1);
+    } else {
+      ++i;
+    }
+  }
+  this.watchStates = [];
+  this.watchListeners = [];
+
   this.picture.addEventListener('change', this.onStorageChange);
+  for (i = 0; i < this.watchStorages.length; ++i) {
+    this.watchListeners[i] = onWatchStorageChange.bind(this,
+                                                       this.watchStorages[i]);
+    this.watchStates[i] = null;
+    this.watchStorages[i].addEventListener('change', this.watchListeners[i]);
+  }
+
   this.emit('volumechanged',{
     video: this.video,
     picture: this.picture
   });
+
+  function onWatchStorageChange(storage, e) {
+    // Uses self to please lint
+    var u = self.findWatch(storage);
+    var value = e.reason;
+    debug('watch[%d] state change: %s', u, value);
+    if (u < 0) { return; }
+
+    var oldValue = self.watchStates[u];
+    self.watchStates[u] = value;
+    if (oldValue === 'shared' || value === 'shared') {
+      self.setState();
+    }
+  }
```

Comment:
I think you have to because you're inside `onWatchStorageChange()` which would have a different `this`.